step3 구현

anjaemin opened the pull request:
- ApiQnaController 추가 및 구현
- ApiQnaAcceptanceTest 추가
- QuestionDTO 수정
- Question domain @JsonIdentityInfo 추가

anjaemin committed:
step3 구현
- ApiQnaController 추가 및 구현
- ApiQnaAcceptanceTest 추가
- QuestionDTO 수정
- Question domain @JsonIdentityInfo 추가

jojoldu commented on src/main/java/nextstep/dto/QuestionDTO.java:
```diff
@@ -3,14 +3,24 @@
 public class QuestionDTO {
-    private final Question question;
-    private final int answerSize;
+    private Question question;
```
아래 컨트롤러 코드를 보면 아마도 이 Dto는 ReponseDto의 역할을 할것 같습니다.
헌데, Reponse에서는 **엔티티를 필드로 쓰면 안됩니다**

수업시간에 말씀드렸던 것처럼 이렇게 사용할 경우 에러가 발생합니다.
Question과 Answer의 필드와 똑같은 필드를 가진 Dto를 반환해야지
Entity를 가진채로 반환하면 어차피 Entity가 Json으로 변환해야되서 안됩니다.

아래와 같이 메인과 테스트에 open in view 옵션을 넣고 수행해보시면 아마 에러가 발생할거에요!

![image](https://user-images.githubusercontent.com/7760496/55682398-dc123980-596d-11e9-9c79-2dc9f92c5211.png)

보통은 엔티티를 그대로 쓰진 않고,
**실제 요청에 필요한 필드들만 가진 Dto를 만들어서 사용합니다**

지금하신것처럼 하면 **만들땐 편하지만 유지보수하기가 너무 힘듭니다**
화면에 필요한 필드가 추가되면 Entity를 변경할순 없는데 어떻게 해야할까요?
OpenInView 옵션을 쓸수도 없습니다.

한번 옵션을 넣어보시고, 테스트해보신뒤 수정을 해보세요 :)

jojoldu reviewed (changes requested):
@anjaemin 재민님 주말에도 고생하셨습니다!

이번엔 좀 큰 주제 2개를 포함해서 코멘트를 남겼습니다 :)
조금만더 하시면 끝나실것 같습니다 :)

anjaemin committed:
step3 피드백 반영
- open-in-view 옵션 false로 변경
- DTO 추가 및 수정
- QnaService 수정(DTO 리턴)
- QnaAcceptaceTest 에서 @Before 부분에 api호출로 질문, 답변 생성
- @After 에서 deleteAll() 메소드 호출

jojoldu commented on src/main/java/nextstep/dto/AnswerDTO.java:
```diff
@@ -0,0 +1,38 @@
+    public AnswerDTO() {
+    }
+
+    public AnswerDTO(Answer answer) {
```
아주 좋습니다 💯

jojoldu commented on src/main/java/nextstep/dto/QuestionDTO.java:
```diff
@@ -1,31 +1,59 @@
-    public QuestionDTO(Question question, int answerSize) {
-        this.question = question;
-        this.answerSize = answerSize;
+    public QuestionDTO(long id, String title, String contents, UserDTO writer, List<AnswerDTO> answers, boolean deleted) {
```
AnswerDto와 마찬가지로 Question 엔티티를 받아서 question.getTitle() 등으로 하는게 더 편하지 않을까요?

jojoldu commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -75,26 +77,34 @@ public void deleteQuestion(User loginUser, long questionId) {
         Question question = findQuestionWithAnswer(questionId);
-        int answerSize = question.getAnswers().size();
-        return new QuestionDTO(question, answerSize);
+        return new QuestionDTO(question.getId(),
```
이 부분이 Answer와 달리 불편하셨을것 같습니다 :)
Answer와 같은 방법으로 해결하시면 될것 같습니다.

jojoldu commented on src/main/resources/application.properties:
```diff
@@ -12,4 +12,6 @@ spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
\ No newline at end of file
+server.servlet.session.tracking-modes=cookie
+
+spring.data.neo4j.open-in-view=false
```
앗 이 옵션이 아닙니다.
jpa.open-in-view 입니다. 
neo4j를 쓰셨네요 :)
이전 PR에 남긴 사진을 다시 한번 참고해보시면 좋을것 같습니다 :)

jojoldu commented on src/test/java/nextstep/web/ApiQnaAcceptanceTest.java:
```diff
@@ -99,7 +105,7 @@ public void addAnswer() {
 
         ResponseEntity<QuestionDTO> questionDTO = template().getForEntity(createLocation, QuestionDTO.class);
-        softly.assertThat(questionDTO.getBody().getQuestion().getAnswers().get(0).getContents()).isEqualTo("answer test");
+        softly.assertThat(questionDTO.getBody().getAnswers().get(0).getContents()).isEqualTo("answer test");
```
.getAnswers().get(0)의 경우는 생각보다 자주 사용되는 경우라서 Dto 내부에 get()을 추가로 만들면 좋을것 같습니다.

```java
QuestionDTO {

  AnswerDto get(int index) {
      return answers.get(index);
  }
}
```

jojoldu commented on src/test/java/nextstep/web/QnaAcceptanceTest.java:
```diff
@@ -124,11 +144,14 @@ public void deleteAnswer() {
         softly.assertThat(response.getHeaders().getLocation().getPath()).startsWith("/questions");
     }
+
+    @After
```
@After 잘 적용해주셨습니다 :)
순서만 변경하시면 될것 같은데요.
@Before 메소드드 다음에 있는게 유지보수하기 용이합니다 :)

jojoldu reviewed (approved):
@anjaemin 재민님 새벽까지 고생하셨어요!
현재 남긴 코드리뷰는 중요한 내용들은 아니니 **참고해서 Step4에 반영**해주세요.
Step3는 정말 잘 구현해주신것 같습니다 :)

수고하셨습니다!
